sts_jax/structural_time_series/learning.py

```diff
@@ -0,0 +1,169 @@
+        M (int): number of fixed samples from q used in evaluation of ELBO.
+
+    Returns:
+        Samples from the approximate posterior q
+    """
+    key1, key2 = jr.split(key, 2)
+    # Make sure the emissions and covariates have batch dimensions
+    batch_emissions = ensure_array_has_batch_dim(emissions, model.emission_shape)
+    batch_inputs = ensure_array_has_batch_dim(inputs, model.inputs_shape)
+
+    initial_unc_params = to_unconstrained(initial_params, param_props)
+
+    @jit
+    def unnorm_log_pos(_unc_params):
+        params = from_unconstrained(_unc_params, param_props)
+        log_det_jac = log_det_jac_constrain(params, param_props)
+        log_pri = model.log_prior(params) + log_det_jac
+        batch_lls = vmap(partial(model.marginal_log_prob, params))(batch_emissions, batch_inputs)
+        lp = batch_lls.sum() + log_pri
+        return lp
+
+    @jit
+    def elbo(vi_hyper, key):
+        """Evaluate negative ELBO at fixed sample from the approximate distribution q.
+        """
+        keys = iter(jr.split(key, 10))
+        # Turn VI parameters and fixed noises into samples of unconstrained parameters of q.
+        unc_params = tree_map(lambda mu, ls: mu + jnp.exp(ls) * jr.normal(next(keys), ls.shape).sum(),
+                              vi_hyper['mu'], vi_hyper['log_sig'])
+        log_probs = unnorm_log_pos(unc_params)
+        log_q = jnp.array(tree_leaves(tree_map(lambda x, *p: norm.logpdf(x, p[0], jnp.exp(p[1])).sum(),
+                                               unc_params, vi_hyper['mu'], vi_hyper['log_sig']))).sum()
+        return log_probs - log_q
+
+    loss_fn = lambda vi_hyp, key: -jnp.mean(vmap(partial(elbo, vi_hyp))(jr.split(key, K)))
+
+    # Fit
+    curr_vi_mus = initial_unc_params
+    curr_vi_log_sigs = tree_map(lambda x: jnp.zeros(x.shape), initial_unc_params)
+    curr_vi_hyper = OrderedDict()
+    curr_vi_hyper['mu'] = curr_vi_mus
+    curr_vi_hyper['log_sig'] = curr_vi_log_sigs
+
+    # Optimize
+    opt_state = optimizer.init(curr_vi_hyper)
+    loss_grad_fn = value_and_grad(loss_fn)
+
+    def train_step(carry, key):
+        vi_hyp, opt_state = carry
+        loss, grads = loss_grad_fn(vi_hyp, key)
+        updates, opt_state = optimizer.update(grads, opt_state)
+        vi_hyp = optax.apply_updates(vi_hyp, updates)
+        return (vi_hyp, opt_state), loss
+    # Run the optimizer
+    initial_carry = (curr_vi_hyper, opt_state)
+    (vi_hyp_fitted, opt_state), losses = lax.scan(
+        train_step, initial_carry, jr.split(key1, num_step_iters))
+
+    # Sample from the learned approximate posterior q
+    vi_sample = lambda key: from_unconstrained(
+        tree_map(lambda mu, s: mu + jnp.exp(s)*jr.normal(key, s.shape),
+                 vi_hyp_fitted['mu'], vi_hyp_fitted['log_sig']),
+        param_props)
+    samples = vmap(vi_sample)(jr.split(key2, num_samples))
+
+    return samples, losses
+
+
+def fit_hmc(
+    model: StructuralTimeSeriesSSM,
```

Comment:
instead of passing in a model object, you could pass in a`prior` and `marginal_loglik` functions,
and then your HMC function would be completely generic, I think